sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala

```diff
@@ -183,16 +172,80 @@ object LogicalRDD {
       }
     }
 
+    val logicalPlan = originDataset.logicalPlan
     val optimizedPlan = originDataset.queryExecution.optimizedPlan
     val executedPlan = originDataset.queryExecution.executedPlan
 
+    val (stats, constraints) = rewriteStatsAndConstraints(logicalPlan, optimizedPlan)
+
     LogicalRDD(
       originDataset.logicalPlan.output,
       rdd,
       firstLeafPartitioning(executedPlan.outputPartitioning),
       executedPlan.outputOrdering,
       isStreaming
-    )(originDataset.sparkSession, Some(optimizedPlan.stats), Some(optimizedPlan.constraints))
+    )(originDataset.sparkSession, stats, constraints)
+  }
+
+  private[sql] def buildOutputAssocForRewrite(
+      source: Seq[Attribute],
+      destination: Seq[Attribute]): Option[Map[Attribute, Attribute]] = {
+    // We check the name and type, allowing nullability, exprId, metadata, qualifier be different
+    // E.g. This could happen during optimization phase.
+    val rewrite = source.zip(destination).flatMap { case (attr1, attr2) =>
+      if (attr1.name == attr2.name && attr1.dataType == attr2.dataType) {
+        Some(attr1 -> attr2)
+      } else {
+        None
+      }
+    }.toMap
+
+    if (rewrite.size == source.size) {
+      Some(rewrite)
+    } else {
+      None
+    }
+  }
+
+  private[sql] def rewriteStatsAndConstraints(
+      logicalPlan: LogicalPlan,
+      optimizedPlan: LogicalPlan): (Option[Statistics], Option[ExpressionSet]) = {
+    val rewrite = buildOutputAssocForRewrite(optimizedPlan.output, logicalPlan.output)
+
+    rewrite.map { rw =>
+      val rewrittenStatistics = rewriteStatistics(optimizedPlan.stats, rw)
+      val rewrittenConstraints = rewriteConstraints(optimizedPlan.constraints, rw)
+
+      (Some(rewrittenStatistics), Some(rewrittenConstraints))
+    }.getOrElse {
+      // can't rewrite stats and constraints, give up
+      logWarning("The output columns are expected to the same (for name and type) for output " +
```

Comment:
This is something I want to discuss. I see the difference of exprId which makes sense, but is it possible to see other difference as well? If it's possible, is LogicalRDD instance even safe?

Suppose we are here, should we raise an internal error instead, or it is still a valid case and we log and give up?

Reply:
I am not aware / can't think of places where the output of analyzed plan and optimized plan would differ in terms of number of columns or their order. Or could some internal thing break that? (some metadata column? some internal hidden column?)
Could the data type differ? (because e.g. of optimizing nullability? optimizing tighter types?)

I'm not an expert in this area, so only speculating.

Reply:
Yeah my understanding is that the final output after optimization should be semantically same. Looks like exprId can change, but I'm not sure what else we allow to be changed. 

My feeling is that changes on type or nullability could also break DSv1 sink, if the change is not done with "compatible way". e.g. table in sink has a column nullability set to "false", and optimizer somehow (shouldn't happen but) changes nullability of a column from "false" to "true" (again, shouldn't happen). Same applies to data type as well. I feel column order also matters, but not sure we pretend that the consumer should access column by name.

Reply:
I've talked offline with @sigmod , his answer is

* expr id can change
* nullability can change (true to false, not vice versa)
* data type and order shouldn't change

That said, it seems to be valid to assert here, though I'd hear more voice on this.

Reply:
I guess without stats and constraints (when you cannot build the valid map), the query should still be executed to the same result as they are only for optimizing query. So I think even for the case, we still can run the query instead of failing it.

And, even you fail the query, I'm not sure what users can do to make valid stats and constraints avaiable?

Reply:
If the query is failing, it clearly denotes the internal bug of Spark and we expect users to report this. But I kind of agree that they shouldn't get blocked with this if the latter part still works fine.